### DISCUSSION TITLE
Add drag and drop to relational interfaces when working with pagination

### DISCUSSION THREAD

**Discussion by cdwmhcc** [directus/directus#12818] [Feature Requests]
### Describe the Bug

![image](https://user-images.githubusercontent.com/8030134/163749964-2e45d6c1-44d4-492e-866f-e0a3ed2b0404.png)

### To Reproduce

After upgrading to 9.9.0, o2m added pagination display, but when there is pagination, the sorting function miss.

### What version of Directus are you using?

v9.9.0

### What version of Node.js are you using?

v16.14.0

### What database are you using?

mysql 8.0.24

### What browser are you using?

chrome

### How are you deploying Directus?

running locally

**Comment by Nitwel** [comment]
This is not bug but a limitation of adding pagination. It would definitely be nice to have that in the future but for that we have to figure out a lot of other things first like how to drag items to the next page.

**Reply by cdwmhcc** [reply]
> This is not bug but a limitation of adding pagination. It would definitely be nice to have that in the future but for that we have to figure out a lot of other things first like how to drag items to the next page.

![image](https://user-images.githubusercontent.com/8030134/164251015-9c2ca9ec-1eac-4e3e-83d6-07da7424258d.png)

Why not use the existing solution for the collection of item lists?

@rijkvanzanten @Nitwel

**Comment by StephenGunn** [comment]
I have a lot of relationships that the sort is crucial and not having it is a real bummer. I also have relationships that have over 100 items in them, so the new paginated layout saves me a lot of trouble there. 

Would it be possible to be able to toggle the paginated display? For something like this:

![image](https://user-images.githubusercontent.com/7240548/165360825-62474082-4593-4d5b-996b-0ae58a637cb7.png)

I would much rather not have pagination and have sorting. The last number on each of those items is the sort number. You can see that it is all over the place. And the sort is crucial to the app.

**Reply by rijkvanzanten** [reply]
@StephenGunn Wanna open a PR to add an option to control how big the individual pages are? That way you could set the page size to 200, and keep sorting for these larger data sets

**Reply by StephenGunn** [reply]
I can try, I don't have that much experience in Vue and the experience I have is in Vue 2. Where would the option go? Would it be on the collection edit page or in the data model setup screen?

**Reply by rijkvanzanten** [reply]
The options are defined here:

https://github.com/directus/directus/blob/29f3a857ad34da616317af48423a31567fa37ddc/app/src/interfaces/list-o2m/index.ts#L15

They're then available as props here:

https://github.com/directus/directus/blob/29f3a857ad34da616317af48423a31567fa37ddc/app/src/interfaces/list-o2m/list-o2m.vue#L108-L112

At that point, I believe it might be as simple as replacing the usage of this limit ref:

https://github.com/directus/directus/blob/29f3a857ad34da616317af48423a31567fa37ddc/app/src/interfaces/list-o2m/list-o2m.vue#L151

with the newly defined prop, but some more testing is needed for that 🙂

**Reply by StephenGunn** [reply]
Thanks, that helps a lot. I will try. I will be in the discord posting noob questions.

**Comment by GioKaaja** [comment]
sorry for the dumb question: is there a way to disable the pagination while team is working on making it possible to manage both features? it's just annoying for our team not to have drag and drop functionality because it represents a key piece on how we expose content on our website. Again, sorry if this is too dumb :)

**Reply by StephenGunn** [reply]
I know, I really need it too and I have been too busy to try and learn how to fix it in the codebase myself. I plan on it, but things keep coming up.

**Reply by NilsTessier** [reply]
quick fix i've found is to display the O2M collection in the admin left menu and then do the sorting on the dedicated content page at `/content/**O2M collection**` 
you have to set both the parent collection and the O2M collection to sort on the same field for it to work properly

**Reply by mariusmoep** [reply]
Same here: We heavily rely on the sorting feature of that field but a whole content team is stuck because we cannot sort images anymore. Is there a way to disable or bypass this feature for now for our non-tech staff?

**Reply by benhaynes** [reply]
We're about to release an update to these relational interfaces that gives more control... but for now there will be no manual reordering across pagination. The best we can do is allow customizing the page count.

**Reply by benhaynes** [reply]
It's worth mentioning, that if a feature is needed by a person/company... we welcome PRs (coordinated with our team beforehand) and sponsored engagements.

**Comment by joshualyness** [comment]
For sorting to work on pagination, could you change the drag behaviour so that it follows the mouse, and on hover over a pagination number, it changes page?

**Reply by benhaynes** [reply]
An interesting idea. Could be worth looking into, but this is definitely a more complex flow to build. :)

**Comment by aliemba** [comment]
It would be a very nice quick fix to make an option to control how big a page is to be. Also if it is larger than "page" it could show in a new side window - as the one like editing a repeater or individual items. This list can handle far more items.

**Comment by rijkvanzanten** [comment]
Added the configurable page size in https://github.com/directus/directus/pull/13772 🙂

**Comment by Anoesj** [comment]
Hmm, any progress on adding a different (extended) view for related content, that disables pagination and allows reordering? It's a sum of all small things like these that make me doubt coming back to Directus in future projects.

**Reply by rijkvanzanten** [reply]
That's where the page size configuration comes in! You can set the page size to something larger than the default if you're expecting to work with a larger nested data set.

**Reply by Anoesj** [reply]
That's just a band-aid IMO..

**Reply by rijkvanzanten** [reply]
How so? What would be the non band-aid solution?

**Reply by Nitwel** [reply]
@Anoesj Would love to hear your thoughts on how drag&drop with pagination could work! Keep in mind that just because you prefer drag&drop over pagination, there are many others that require pagination, so we can't just throw pagination out of the window.

The main problem with pagination and drag&drop is that:
1. How do you drag an item to the next page?
2. How do you tell the api what needs to be dragged where?

Especially when dragging to a new page, how do we make sure that previews for items which are dragged onto another page still get loaded as the api doesn't know about the drag&drop changes yet. So you could basically end up with a page that has 14 and the next that has 16 which is not good at all.

**Reply by Anoesj** [reply]
How about this: when related content is paginated and cannot be sorted anymore, display a "View all and sort" option. This opens a new overview page with all related items in one big view, where all items are loaded and draggable. This would preserve the pagination feature but also add a solution for this weird UX edge case where sorting is not possible anymore because of pagination, but it's not clear to end users why that is and what to do about it.

**Reply by Nitwel** [reply]
It is not feasable to add a button that sets the limit to infinity as this could litterally dos your api but we could add a button that you can opt-in that basically says `increase page size by X` which does exacly what it says. So if you have a relation with 70 items and your page size is 50, you can click that button to load the other 20 items and then we can enable drag&drop again. What do you think Rijk?

In the case that Rijk also things this a fair in between, mind opening a PR for that feature?

**Reply by rijkvanzanten** [reply]
> It is not feasable to add a button that sets the limit to infinity as this could litterally dos your api but we could add a button that you can opt-in that basically says `increase page size by X` which does exacly what it says. So if you have a relation with 70 items and your page size is 50, you can click that button to load the other 20 items and then we can enable drag&drop again. What do you think Rijk?
> 
> 
> 
> In the case that Rijk also things this a fair in between, mind opening a PR for that feature?

Isn't that basically the same result as setting the page size to a number bigger than your expected use? 🤔

**Reply by Anoesj** [reply]
Why would it DoS the API? You can load all related items in one call, right?

**Reply by paescuj** [reply]
> Why would it DoS the API? You can load all related items in one call, right?

Since the number of items can be an arbitrary high and depending on the content further relations have to be fetched!

**Reply by Anoesj** [reply]
Wouldn't all items have to be loaded anyway before a user is able to change the items' order using drag and drop? Increasing the page size to a huge number would result in the same high number of API calls, but every time the item's edit page is accessed, instead of only when the user clicks the "View all and sort" button. So it seems to me that the DoS argument doesn't hold up, since the same problem would arise now by increasing the page size.

**Reply by rijkvanzanten** [reply]
> Wouldn't all items have to be loaded anyway before a user is able to change the items' order using drag and drop? 

Yes! This is why there's pagination in the first place 🙂 Right now, drag and drop is available if all items are present in a page. When building this functionality, in our research, it turned out that a very very small fraction of the user base was doing manual sorting on nested datasets > 100 items, so we opted for the current system, as it offers the correct balance between maintenance overhead and feature complexity for the 80/20. You can increase the page-size to something that fits your use case, but because there's an upper limit, you don't end up breaking your setup.

**Reply by Anoesj** [reply]
I get it, but I don't think it's good UX/DX. It should be either clearer for the end user why the view cannot be sorted, or a solution should be offered on the spot, like a "View all and sort" button. I still think a "View all and sort" button would not create any DoS problems any more than increasing the page size would. Is this feature really that much maintenance overhead/complexity?

**Reply by rijkvanzanten** [reply]
>  I still think a "View all and sort" button would not create any DoS problems any more than increasing the page size would.

I don't think we can ever go with a "view all".. Keep in mind that there's scenarios where you can have millions of related items, at which point you'll run into serious problems. We could do a "load next 25" or something though! A dropdown to allow the user to change the page size on the fly would also be a nice enhancement. 

> Is this feature really that much maintenance overhead/complexity?

The complexity comes in when you do drag and drop across pages. Doing drag and drop within a single page is what we currently support already 👍🏻

**Reply by Anoesj** [reply]
Sound like a fine solution then :smile:

**Comment by paescuj** [comment]
On the weekend I thought about how sorting over several pages could look like, in case we want to tackle this in the future:

Assuming that it's usually not known beforehand at which position the element is to be repositioned, it would be convenient if the entire list of items could still be explored during the drag and drop process.
At the same time we cannot and don't want to load all items at once, so we either need a mechanisms to load more items (load more button) or rely on pagination for this as well.
With that in mind, this was my idea of how things could work:
1. Grab an item and drag it over the pagination bar / buttons
2. A separate "sort interface" opens up (for example in the drawer or in a modal)
3. The dragged item to sort is pinned on top, otherwise the same interface, incl. pagination
    (inside the list the pinned item could be highlighted)
4. Now browse through the pages via pagination and once the desired position is found, drag and drop the pinned item from above.

![paginated-dnd(1)](https://user-images.githubusercontent.com/5363448/214079277-a2c454a1-d478-4d29-a592-74de6ad4f6a2.png)

**Reply by lukas-schaetzle** [reply]
Great suggestion. Just wanted to propose a similar idea :)
However, I don't see why a modal or drawer is necessary for that when we could just use the same interface for it. I would just fade in the seperated region to pin an element when an element is being dragged (and pagination is used).

**Reply by paescuj** [reply]
Sure, that would be possible as well 👍

**Reply by benhaynes** [reply]
Really thoughtful UX, @paescuj... I like it! Could be in a drawer or the same listing page, but the benefit of the drawer is it's Layout agnostic... so we could have easier temporary sorting on a Map layout, for example. 👍

**Reply by paescuj** [reply]
Thanks! ❤️ 

> [...] but the benefit of the drawer is it's Layout agnostic... so we could have easier temporary sorting on a Map layout, for example.

That's a good point 👍

**Reply by rijkvanzanten** [reply]
@paescuj This is where the fun begins!

Lets say pages are 10 items each, and you navigate to page 5. This means that you're fetching items with offset = 5 * 10. If you then move an item from page 5 to page 4, how are you gonna keep track of the offsets of the pages? Now page 4 should be offset 4 * 10, but with a page size of 9 to keep the overall page at 10. If you navigate back to page 5, page 5 should now be offset 5 * 10 + 1, as an item from 5 was moved to 4. If you do this `n` times, things get complicated very very quickly 😬

**Reply by lukas-schaetzle** [reply]
I think page size should always stay the same. When you drop the item into its new place on another page, the page limit is exceeded by 1. So the first or last item has to move to the preceding or following page, depending on whether the new page in which the item is placed is before or after the old page position. I would indicate this by fading out and maybe a sliding animation for the moving item after dropping the pinned item.
If we go to another page after the item has been moved we just reload the items.

**Reply by cdwmhcc** [reply]
Good idea! But this solution needs to solve a problem: the impact on the sorting of tens of thousands or even hundreds of thousands of data.

**Reply by benhaynes** [reply]
I like the UX of the above solution, I think the bigger issue is whether we can avoid updating the `sort` column value for all items between the sort start and start end records. If we move an item from sort position `5` to `5000`, then we would need to update the sort values for the 4,999 items in between.

Any thoughts around how we could handle that?

In the past we were thinking we could sort by 100s, (eg: 100, 200, 300)... so we could insert items "between" without updating the other item values. But eventually, you run out of "room" between, and it just seems messy. Any other ideas?

**Reply by lukas-schaetzle** [reply]
> But eventually, you run out of "room" between, and it just seems messy. Any other ideas?

You could use float values for the sort values. Inserting at the start or end would just normally increase or decrease the value by 1. When you insert an item between two items, its sort value is set to the average sort value between the preceding and following item. E.g. insert between values 1 and 2 -> 1.5, between 1 and 1.5 -> 1.25 and so on.
When the sort value is saved as a string in the DB we don't run into the problem that float values can only be saved with a fixed precision and so we could therotically insert an unlimited number of times without running out of numbers.

However this solution comes with some problems:
1. The sort value may look really akward after many insertions which is not pleasant for developers who want to configure the value manually.
2. We have a problem, when both neighboring items have the same sort value.

**Edit:**
Actually, it would make more sense to just append another decimal digit instead of calculating the average. That way the number of digits does not increase as fast. Example:
- insert between 1 and 2 -> 1.1
- insert between 1.1 and 2 -> 1.2
- insert between 1 and 1.1 -> 1.01

**Reply by cdwmhcc** [reply]
I don't think this is a perfect solution. But I couldn't find a better solution than this either.

**Reply by benhaynes** [reply]
I agree. It's not a perfect solution, it would work (I think) but I'm a bit of a perfectionist, and that data after a bunch of sorting seems smelly.

**Reply by Nitwel** [reply]
Funny to see that you're all basically going through the same thought process I went when doing the relational rework.

That said, keep in mind that we load only single pages, so if you are on page 7 of 10, you only know the items that the **API** thinks should be on page 7. So if you would move an item from page 6 to 7, we can't just display it in page 7 because the api still thinks the data should be as is. The only possible way I see this working is that after each drag&drop, we update the items in the api, but that's not possible because all changes should only be saved when you click the save icon at the top of the item view.

**Reply by paescuj** [reply]
Thanks, you've all made very valid and good points! 👍 Sorting is really a tough topic!

I imagined that we could take a very simple approach in a first step and then revise the whole thing at a later stage once a suitable solution has been found.
At the moment there is no possibility at all to sort over several pages. So if we could provide at least a simple solution, I think that would already be very valuable.

Referring again to my example above, I envisioned a flow like this:

1. Drawer with the "sort interface" opens up.
2. User navigates to the desired page, drags the item from the top and drops it at a specific position.
    The superfluous item in the list will be hidden (but not yet removed) and the pagination is locked.
    Now the user has two possibilities:
    - **Reset**: The dropped item will be moved and pinned to the top again and the list is reset (superfluous item is shown again and pagination unlocked). The user goes back to step 2.
    - **Save**: The new position of the item is updated (probably via [sort API call](https://docs.directus.io/reference/system/utilities.html#manually-sort-items-in-collection)), the drawer gets closed and the current page of items in the "normal" view is re-fetched.
      
      ---
       The only problem here, is that it deviates somewhat from the normal saving procedure, as mentioned by @Nitwel:
       > [...] but that's not possible because all changes should only be saved when you click the save icon at the top of the item view.
      
       However, since it would happen separately in the drawer and the user would explicitly press a save button, it might be acceptable...
    
Note: I don't not want to impose this solution at all, I just want to show that it might be possible to come up with a simpler solution in the short term, so that the users would at least have a possibility to do some sorting.

**Reply by rijkvanzanten** [reply]
> Note: I don't not want to impose this solution at all, I just want to show that it might be possible to come up with a simpler solution in the short term, so that the users would at least have a possibility to do some sorting.

I think a good starting point is informing the user why they can't sort. After that, adding the ability to increase the page size on the fly when using the interface should allow the end-user to manually re-enable sorting

**Reply by paescuj** [reply]
> I think a good starting point is informing the user why they can't sort. After that, adding the ability to increase the page size on the fly when using the interface should allow the end-user to manually re-enable sorting

Sounds like a plan 👍

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
